File: python/Bit Manipulation/single_number.py
from typing import List

class Solution:
    def singleNumber(self, nums: List[int]) -> int:
        """
        Approach:
        - Use a hash map to count occurrences of each number.
        - Return the number that appears only once.

        Time Complexity: O(n)
        Space Complexity: O(n)
        """
        numCount = {}
        # Count occurrences of each number
        for num in nums:
            numCount[num] = numCount.get(num, 0) + 1
        
        # Find the number that appears only once
        for num, count in numCount.items():
            if count == 1:
                return num
        
        return 0  # or appropriate default value if no single number found
    
class Solution2:
    def singleNumber(self, nums: List[int]) -> int:
        """
        Approach:
        - Use XOR to find the single number. All pairs cancel out, leaving the unique number.

        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        res = 0
        for n in nums:
            res = n ^ res
        return res

# -------------------
# Test Cases
# -------------------
def test_single_number():
    s1 = Solution()
    s2 = Solution2()
    # Test case 1: Example from Leetcode
    nums1 = [2,2,1]
    assert s1.singleNumber(nums1) == 1
    assert s2.singleNumber(nums1) == 1
    # Test case 2: Single element
    nums2 = [42]
    assert s1.singleNumber(nums2) == 42
    assert s2.singleNumber(nums2) == 42
    # Test case 3: Negative numbers
    nums3 = [-1, -1, -2]
    assert s1.singleNumber(nums3) == -2
    assert s2.singleNumber(nums3) == -2
    # Test case 4: Larger array
    nums4 = [4,1,2,1,2]
    assert s1.singleNumber(nums4) == 4
    assert s2.singleNumber(nums4) == 4
    # Test case 5: All numbers are the same except one
    nums5 = [7,7,7,8,7]
    assert s1.singleNumber(nums5) == 8
    assert s2.singleNumber(nums5) == 8
    print("All test cases passed!")

if __name__ == "__main__":
    test_single_number()

File: C++/Bit Manipulation/single_number.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int singleNumber(vector<int>& nums) {
        
    }
};
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    /*
     * Approach:
     * - Use XOR (^) to find the single number.
     * - Every number that appears twice cancels out (since n ^ n = 0).
     * - The number that appears only once remains at the end.
     *
     * Time Complexity: O(n)
     * - We iterate through the array once.
     *
     * Space Complexity: O(1)
     * - We only use a single integer variable for the result.
     */
    int singleNumber(vector<int>& nums) {
        int res = 0;
        for (int n : nums) {
            res ^= n;  // XOR current number with result
        }
        return res;
    }
};

/*
 * 📖 Line-by-line explanation:
 *
 * #include <bits/stdc++.h>        → Includes all standard C++ libraries.
 * using namespace std;            → Allows usage of standard library names without `std::` prefix.
 *
 * class Solution { … };           → Defines a class named Solution.
 *
 * public:                         → All members below are accessible outside the class.
 *
 * int singleNumber(vector<int>& nums) { … } → Function that finds the single number.
 *
 * int res = 0;                    → Initialize result variable to 0.
 *
 * for (int n : nums) { … }       → Range-based for loop to iterate over all elements in nums.
 *
 * res ^= n;                       → XOR current number with res. (res = res ^ n)
 *                                  → Numbers appearing twice cancel out, leaving the unique number.
 *
 * return res;                      → Return the single number.
 *
 */
